Домашняя работа 6

khitao opened the pull request:
1) Реализовал две транзакции для сохранения данных из внешнего сервиса в базу данных, используя разные уровни изоляции:

- Первая транзакция управляется через JDBC.

- Вторая транзакция управляется с использованием Spring.

2) Создал индексы с помощью Liquibase:

- Добавил индекс по одному полю.

- Создал составной индекс.

3) Добавил новые методы в сервисы для контроля количества определенных объектов в базе данных.

4) Реализовал кастомное исключение для обработки ситуаций, когда пытаются добавить уже существующий город или погоду в базу данных.

khitao committed:
Домашняя работа 6

1) Реализовал две транзакции для сохранения данных из внешнего сервиса в базу данных, используя разные уровни изоляции:

- Первая транзакция управляется через JDBC.

- Вторая транзакция управляется с использованием Spring.

2) Создал индексы с помощью Liquibase:

- Добавил индекс по одному полю.

- Создал составной индекс.

3) Добавил новые методы в сервисы для контроля количества определенных объектов в базе данных.

4) Реализовал кастомное исключение для обработки ситуаций, когда пытаются добавить уже существующий город или погоду в базу данных.

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/service/WeatherApiJpaTransactionService.java:
```diff
@@ -0,0 +1,72 @@
+        this.weatherTypeService = weatherTypeService;
+    }
+
+    @Transactional(isolation = Isolation.REPEATABLE_READ, rollbackFor = EntityAlreadyExistsException.class)
```
почему Rollback только для этого Exception-а? А если стрельнёт, что-то ещё откатывать не будем?

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/SpringBootAppApplication.java:
```diff
@@ -10,4 +16,21 @@ public static void main(String[] args) {
     }
 
+    @Bean
+    public CommandLineRunner commandLineRunner(WeatherApiJdbcTransactionService weatherApiJdbcTransactionService,
+                                               WeatherApiJpaTransactionService weatherApiJpaTransactionService,
+                                               WeatherApiService weatherApiService) {
+        return args -> {
+
+            WeatherApi weatherApi = weatherApiService.getCurrentWeather("London").block();
+
+            weatherApiJdbcTransactionService.addDataFromWeatherApi(weatherApi);
+
+            weatherApi = weatherApiService.getCurrentWeather("London").block();
+
+            weatherApiJpaTransactionService.addDataFromWeatherApi(weatherApi);
+
+        };
+    }
```
В задаче сказано: 

`Полученные данные по ресту из внешнего сервиса (Лекция 4) сохранить в бд.`

Нужно было добавить сохранение в БД в момент запроса по REST-у `/weather_api/current/{location}`

madoed commented on SpringBootApp/src/main/java/ru/khodov/springbootapp/service/WeatherApiJdbcTransactionService.java:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    public void addDataFromWeatherApi(WeatherApi weatherApi) {
+        DefaultTransactionDefinition def = new DefaultTransactionDefinition();
+        def.setIsolationLevel(TransactionDefinition.ISOLATION_SERIALIZABLE);
+        TransactionStatus transaction = transactionManager.getTransaction(def);
```
Это не управление через JDBC транзакцией, это управление через Spring в ручном режиме 

https://www.tutorialspoint.com/jdbc/jdbc-transactions.htm

madoed reviewed (changes requested):
Доделать по замечаниям

khitao committed:
Исправление Домашняя работа 6

Убрал из аннотации @Transactional ненужный там Rollback

Добавил сохранение в БД в момент запроса по REST-у /weather_api/current/{location}

Переделал на управление через JDBC транзакцией

Добавил интерфейс для двух реализаций транзакций
